# Problem 1.5.1

An 11-symbol message has been encrypted as follows. Each symbol is represented by a number between 0 and 26 (A -> 0, B -> 1, ..., Z -> 25, space -> 26). Each number
is represented by a five-bit binary sequence (0 -> 00000, 1 -> 00001, ...,26 -> 11010). Finally, the resulting sequence of 55 bits is encrypted using a flawed version of the one-time pad: the
key is not 55 random bits but 11 copies of the same sequence of 5 random bits. The cyphertext is

10101 00100 10101 01011 11001 00011 01011 10101 00100 11001 11010

Try to find the plaintext.

In [1]:
import string

### 1. Create Mappings from Upper Case Letters to Binary Strings and Vice Versa

In [2]:
symbols = string.ascii_uppercase + ' '
alpha_to_bin = {s: format(n, "05b") for n, s in enumerate(symbols)} 
bin_to_alpha = {v:k for k, v in alpha_to_bin.items()}

In [3]:
alpha_to_bin

{'A': '00000',
 'B': '00001',
 'C': '00010',
 'D': '00011',
 'E': '00100',
 'F': '00101',
 'G': '00110',
 'H': '00111',
 'I': '01000',
 'J': '01001',
 'K': '01010',
 'L': '01011',
 'M': '01100',
 'N': '01101',
 'O': '01110',
 'P': '01111',
 'Q': '10000',
 'R': '10001',
 'S': '10010',
 'T': '10011',
 'U': '10100',
 'V': '10101',
 'W': '10110',
 'X': '10111',
 'Y': '11000',
 'Z': '11001',
 ' ': '11010'}

In [4]:
bin_to_alpha

{'00000': 'A',
 '00001': 'B',
 '00010': 'C',
 '00011': 'D',
 '00100': 'E',
 '00101': 'F',
 '00110': 'G',
 '00111': 'H',
 '01000': 'I',
 '01001': 'J',
 '01010': 'K',
 '01011': 'L',
 '01100': 'M',
 '01101': 'N',
 '01110': 'O',
 '01111': 'P',
 '10000': 'Q',
 '10001': 'R',
 '10010': 'S',
 '10011': 'T',
 '10100': 'U',
 '10101': 'V',
 '10110': 'W',
 '10111': 'X',
 '11000': 'Y',
 '11001': 'Z',
 '11010': ' '}

### 2. The Cipher

In [5]:
cipher = ['10101', '00100', '10101', '01011', '11001', '00011', '01011', '10101', '00100', '11001', '11010']

### 3. The Keys

Since there can be a maximum of 32 keys, generate and save them in a list

In [6]:
keyList = [format(i, "05b") for i in range(32)]

### 4. Decrypt Function

In [7]:
def decrypt(kList, cipher):
    """
    Given a list of possible 5-bit keys and a cipher of 5-bit symbols, returns a
    dictionary of keys and decrypted messages.
    
    """
    
    d = {}
    for k in kList:
        afterXor = [(int(c, 2) ^ int(k, 2)) % 27 for c in cipher]
        afterXor_to_bin = [format(n, "05b") for n in afterXor]
        decrypted = [bin_to_alpha[k] for k in afterXor_to_bin]
        d[k] = decrypted
    return d

In [8]:
d= decrypt(keyList, cipher)

In [9]:
for k, v in d.items():
    print(k + " -> " + "".join(v))

00000 -> VEVLZDLVEZ 
00001 -> UFUKYCKUFYA
00010 -> XGXJABJXGAY
00011 -> WHWI AIWH Z
00100 -> RARPCHPRACD
00101 -> QBQOBGOQBBE
00110 -> TCTNEFNTCEB
00111 -> SDSMDEMSDDC
01000 -> CMCDRLDCMRS
01001 -> BNBCQKCBNQT
01010 -> EOEBTJBEOTQ
01011 -> DPDASIADPSR
01100 -> ZIZHVPHZIVW
01101 -> YJYGUOGYJUX
01110 -> AKAFXNFAKXU
01111 ->  L EWME LWV
10000 -> FUFAJTAFUJK
10001 -> EVE IS EVIL
10010 -> HWHZLRZHWLI
10011 -> GXGYKQYGXKJ
10100 -> BQBENXEBQNO
10101 -> ARADMWDARMP
10110 -> DSDCPVCDSPM
10111 -> CTCBOUBCTON
11000 -> NBNTBATNBBC
11001 -> MCMSA SMCAD
11010 -> PDPRDZRPDDA
11011 -> OEOQCYQOECB
11100 -> JYJXFEXJYFG
11101 -> IZIWEDWIZEH
11110 -> L LVHCVL HE
11111 -> KAKUGBUKAGF


### 5. The Plaintext

A scan of the decrypted messages above shows one that reads "EVE IS EVIL". The associated key is "10001".